m3/src/reporter_impl.h

```diff
@@ -0,0 +1,146 @@
+
+  // Ensure the class is non-copyable.
+  Impl(const Impl &) = delete;
+
+  Impl &operator=(const Impl &) = delete;
+
+  // Methods to implement the StatsReporter interface.
+  std::unique_ptr<tally::Capabilities> Capabilities();
+
+  void Flush();
+
+  void ReportCounter(const std::string &name,
+                     const std::map<std::string, std::string> &tags,
+                     int64_t value);
+
+  void ReportGauge(const std::string &name,
+                   const std::map<std::string, std::string> &tags,
+                   double value);
+
+  void ReportTimer(const std::string &name,
+                   const std::map<std::string, std::string> &tags,
+                   std::chrono::nanoseconds value);
+
+  void ReportHistogramValueSamples(
+      const std::string &name, const std::map<std::string, std::string> &tags,
+      uint64_t bucket_id, uint64_t num_buckets, double buckets_lower_bound,
+      double buckets_upper_bound, uint64_t samples);
+
+  void ReportHistogramDurationSamples(
+      const std::string &name, const std::map<std::string, std::string> &tags,
+      uint64_t bucket_id, uint64_t num_buckets,
+      std::chrono::nanoseconds buckets_lower_bound,
+      std::chrono::nanoseconds buckets_upper_bound, uint64_t samples);
+
+ private:
+  // Run implements the logic of the Reporter, pulling metrics of its queue and
+  // emitting them.
+  void Run();
+
+  // Enqueue adds a metric to the Reporter's queue.
+  void Enqueue(thrift::Metric);
+
+  // Process processes a metric so it can be emitted.
+  void Process(thrift::Metric metric,
+               std::shared_ptr<TCalcTransport> calc_transport);
+
+  // Helper methods used when processing metrics.
+  void ReportMetric(const std::string &name,
+                    const std::set<thrift::MetricTag> tags,
+                    thrift::MetricValue value);
+
+  thrift::MetricValue CreateCounter(int64_t value);
+
+  thrift::MetricValue CreateGauge(double value);
+
+  thrift::MetricValue CreateTimer(std::chrono::nanoseconds value);
+
+  std::set<thrift::MetricTag> ConvertTags(
+      const std::map<std::string, std::string> &tags);
+
+  std::string ValueBucketString(double bucket_bound);
+
+  std::string DurationBucketString(std::chrono::nanoseconds bucket_bound);
+
+  std::string BucketID(uint64_t bucket_id, uint64_t num_buckets);
+
+  std::string FormatDuration(std::chrono::nanoseconds duration, int precision);
+
+  static const char HISTOGRAM_BUCKET_NAME[];
+  static const char HISTOGRAM_BUCKET_ID_NAME[];
+
+  const std::set<thrift::MetricTag> common_tags_;
+  const uint32_t max_queue_size_;
+  const uint16_t max_packet_size_;
+
+  std::condition_variable cv_;
+  std::thread thread_;
+  std::shared_ptr<TUDPTransport> transport_;
+
+  std::mutex run_mutex_;
+  bool run_;
+
+  std::mutex emission_mutex_;
+  thrift::MetricBatch emission_batch_;
+  std::unique_ptr<thrift::M3Client> emission_client_;
+
+  std::mutex queue_mutex_;
+  std::queue<thrift::Metric> queue_;
```

Comment:
Ah! Don't ever use queue. It is a container adapter and not an actual container. Use `deque`.

Reply:
Hmm, we use it entirely as a FIFO queue in this case so wouldn't it be preferable to use the more restrictive queue API?

Reply:
Actually I believe container adapters are just API bridges for other containers. For example, `std::queue` allows you to treat `std::vector` like a queue. It does not help your performance and probably hurts it if you use a bad underlying type. In this case, since you didn't specify the underlying container, it is using `std::deque` by default. See more here: https://en.cppreference.com/w/cpp/container/queue.

Reply:
I'd suggest `std::vector` - it guarantees contiguous memory, which will give you better locality when draining the queue.

Reply:
I think `deque` is fine, according to the [documentation](http://www.cplusplus.com/reference/deque/deque/):

> Therefore, deques are a little more complex internally than vectors, but this allows them to grow more efficiently under certain circumstances, especially with very long sequences, where reallocations become more expensive.